src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,78 +34,236 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
     }
 
     private void processInput(String userInput) {
+        filterMilestonesByQuerySentence(allMilestones, userInput);
+        determineBestMatchingMilestones(userInput);
         if (userInput.isEmpty()) return;
+        toggleFirstMilestone(bestMatchingMilestones);
+    }
 
-        String[] userInputWords = userInput.split(" ");
+    private void filterMilestonesByQuerySentence(List<PickerMilestone> milestones, String querySentence) {
+        String[] userInputWords = querySentence.split(" ");
         Stream.of(userInputWords)
-                .forEach(this::filterMilestones);
+                .forEach(queryWord -> filterMilestonesByQueryWord(milestones, queryWord));
+    }
 
-        toggleFirstMatchingMilestone(currentMilestonesList);
+    private void filterMilestonesByQueryWord(List<PickerMilestone> milestones, String queryWord) {
+        milestones.stream()
+                .forEach(milestone -> {
+                    if (milestone.isMatching()) milestone.setMatching(isMatchingQuery(milestone, queryWord));
+                });
     }
 
     /**
-     * Toggles the milestone that has milestoneName as its exact name
-     * <p>
-     * This method is case-sensitive
+     * Toggles the selection status of the milestone that has milestoneName as its exact name
      *
+     * This method is case-sensitive
      * @param milestoneName
      */
     public final void toggleExactMatchMilestone(String milestoneName) {
-        Optional<PickerMilestone> onlyMatchingMilestone = getExactMatchMilestone(currentMilestonesList, milestoneName);
-        onlyMatchingMilestone.ifPresent(milestone -> toggleMilestone(milestone));
+        Optional<PickerMilestone> exactMatchMilestone = getExactMatchMilestone(allMilestones, milestoneName);
+        exactMatchMilestone.ifPresent(this::toggleMilestone);
     }
 
     /**
-     * Toggles the milestone that is the first match to the given query
-     * <p>
-     * This method is NOT case-sensitive
+     * Toggles the selection status of the first milestone in the list
      *
      * @param milestoneList
      */
-    private final void toggleFirstMatchingMilestone(List<PickerMilestone> milestoneList) {
-        Optional<PickerMilestone> firstMatchingMilestone = getFirstMatchingMilestone(milestoneList);
+    private final void toggleFirstMilestone(List<PickerMilestone> milestoneList) {
+        Optional<PickerMilestone> firstMatchingMilestone = milestoneList.stream()
```

Comment:
Variable name is misleading. There is no other mention of matching in this whole method.